vectordraw.js

```diff
@@ -0,0 +1,282 @@
+
+    // If this vector is already rendered, only update its coordinates.
+    var board_object = this.board.elementsByName[vec.name];
+    if (board_object) {
+        board_object.point1.setPosition(JXG.COORDS_BY_USER, coords[0]);
+        board_object.point2.setPosition(JXG.COORDS_BY_USER, coords[1]);
+        return;
+    }
+
+    var tail = this.board.create('point', coords[0], {
+        size: 1,
+        name: vec.name,
+        withLabel: false
+    });
+    var tip = this.board.create('point', coords[1], {
+        size: 2,
+        name: vec.name
+    });
+    var arrow = this.board.create('arrow', [tail, tip], {name: vec.name});
+
+    // Disable the <option> element corresponding to vector.
+    var option = this.element.find('.menu option[value=' + idx + ']');
+    option.prop('disabled', true).prop('selected', false);
+
+    return arrow;
+};
+
+VectorDraw.prototype.removeVector = function(idx) {
+    var vec = this.settings.vectors[idx];
+    var object = this.board.elementsByName[vec.name];
+    if (object) {
+        this.board.removeAncestors(object);
+        // Enable the <option> element corresponding to vector.
+        var option = this.element.find('.menu option[value=' + idx + ']');
+        option.prop('disabled', false);
+    }
+};
+
+VectorDraw.prototype.addVectorFromList = function() {
+    this.pushHistory();
+    var idx = this.element.find('.menu select').val();
+    this.renderVector(idx);
+};
+
+VectorDraw.prototype.reset = function() {
+    this.pushHistory();
+    JXG.JSXGraph.freeBoard(this.board);
+    this.render();
+};
+
+VectorDraw.prototype.pushHistory = function() {
+    this.history_stack.undo.push(this.getState());
+    this.history_stack.redo = [];
+};
+
+VectorDraw.prototype.undo = function() {
+    var state = this.history_stack.undo.pop();
+    if (state) {
+        this.history_stack.redo.push(this.getState());
+        this.setState(state);
+    }
+};
+
+VectorDraw.prototype.redo = function() {
+    var state = this.history_stack.redo.pop();
+    if (state) {
+        this.history_stack.undo.push(this.getState());
+        this.setState(state);
+    }
+};
+
+VectorDraw.prototype.getMouseCoords = function(evt) {
+    var i = evt[JXG.touchProperty] ? 0 : undefined;
+    var c_pos = this.board.getCoordsTopLeftCorner(evt, i);
+    var abs_pos = JXG.getPosition(evt, i);
+    var dx = abs_pos[0] - c_pos[0];
+    var dy = abs_pos[1] - c_pos[1];
+
+    return new JXG.Coords(JXG.COORDS_BY_SCREEN, [dx, dy], this.board);
+};
+
+VectorDraw.prototype.onBoardDown = function(evt) {
+    this.pushHistory();
+
+    var can_create = true;
+    var coords = this.getMouseCoords(evt);
+
+    // Can't create a vector if none is selected from the list.
+    var vec_idx = this.element.find('.menu select').val();
+    if (!vec_idx) {
+        return;
+    }
+
+    // Prevent creating a new vector when dragging tip or tail of existing vector.
+    for (var eid in this.board.objects) {
```

Comment:
@mtyaka nit: in my opinion this loop deserves a dedicated function. It used to be 9-lines long, and now it's 20, so probably he's a grown up now and should find itself a new home :)